### ISSUE TITLE
[ERROR] Bus error (core dumped)

### ISSUE THREAD

**Issue by LeungH** [facebookresearch/PyTorch-BigGraph#33]
My environment：python 3.6.8 ：：anaconda、pytorch 1.0.1.post2、cuda 9.0
I follow the guide
pip install torchbiggraph
examples/livejournal.py
And then i meet an error “Bus error (core dumped)”
Can you tell me how to fix this?

//my terminal's output
Using downloaded and verified file: data/soc-LiveJournal1.txt.gz
Extracting data/soc-LiveJournal1.txt.gz
Downloaded and extracted file.
Shuffling and splitting train/test file. This may take a while.
Reading data from file:  data/soc-LiveJournal1.txt
Shuffling data
Splitting to train and test files
Found some files that indicate that the input data has already been preprocessed, not doing it again.
These files are: data/livejournal/dictionary.json, data/livejournal/entity_count_user_id_0.txt, data/train_partitioned/edges_0_0.h5, data/test_partitioned/edges_0_0.h5
2019-04-16 11:58:14  Loading entity counts...
2019-04-16 11:58:14  Creating workers...
2019-04-16 11:58:14  Initializing global model...
Bus error (core dumped)

**Comment by SergYevtukhoff** [comment]
Same problem when trying to train model on a tiny graph (300K nodes, 1M edges)

**Comment by ggliem** [comment]
Hi there,
had the same problem a few days ago and figured out that it crashed when calling the share_memory method. I was running my code inside a docker container and the following issue helped me resolving the problem:
https://github.com/pytorch/pytorch/issues/2244

Thought this might be worth sharing in case you are running your code inside a docker container too.

**Comment by lerks** [comment]
Sorry for the late reply.

I have never seen that error, so I don't really know how to help. Since it doesn't produce a traceback it must be happening deep inside some C-level code, most likely inside PyTorch. To figure this out properly, it would help to manually trace the execution (that is, using a debugger like `pdb` to step into functions, _manually_ keeping track of the stack) until the error occurs.

However, the findings of @ggliem (thanks a lot for sharing them here!) seem to me to be a very plausible explanation. PBG does use shared memory for the model and for all the embeddings all the time, no matter whether it's running with only just worker process, in distributed mode, with dynamic relations, etc.

@LeungH, @SergYevtukhoff: can you try to see whether this helps you solve the problem? The linked PyTorch issue has some tips in case you're using Docker, but there could be other constraints on shared memory coming from elsewhere. As I'm not an expert, the best suggestion I can give is to google that...

**Comment by adamlerer** [comment]
Sorry also for the late reply. PBG fails with a "Bus error" when it runs out of shared memory (PBG has to put all the embeddings in shared memory for multi-threaded training). Unfortunately this error is produced from the kernel and there's really no way to catch it or detect it early.

The possible solutions are:
1) You can try increasing shared memory. If you're running in docker then you may have very little shm by default and pytorch/pytorch#2244 is relevant. If you're just running on a linux machine you can look here https://masukkhan.wordpress.com/2015/12/09/resize-devshm-filesystem-in-linux/
2) You can reduce the amount of memory you need to use, e.g. by reducing `dim` or increasing `num_partitions`.

@lerks this is such a common failure mode we might want to add it to the wiki somewhere.

**Comment by lerks** [comment]
Closing due to inactivity. Please reopen if you still have questions.

PS: in #40 I added a troubleshooting section to the doc that mentions this issue and possible solutions.

**Comment by sademakn** [comment]
Hi, I am getting an error like the reported error only when I'm running my code on docker and it is :

```
pgb_app    | 2021-02-05 08:51:36,353   [Trainer-0] Edge path: data/edges_partitioned
pgb_app    | 2021-02-05 08:51:36,354   [Trainer-0] still in queue: 0
pgb_app    | 2021-02-05 08:51:36,354   [Trainer-0] Swapping partitioned embeddings None ( 0 , 0 )
pgb_app    | 2021-02-05 08:51:36,355   [Trainer-0] Loading partitioned embeddings from checkpoint
pgb_app    | Bus error (core dumped)
pgb_app exited with code 135
```
number of nodes are about 1M .

**Comment by lw** [comment]
As mentioned before, "Bus error" typically indicates an out-of-memory issue. The solution is to use a higher number of partitions, so that each of them is smaller.

**Comment by lw** [comment]
Note that you in addition to checking how much memory your (bare metal) machine has, you should also check how much of that memory is accessible by the Docker container for use as shared memory, which is typically a much smaller amount. For reasons such as these, we recommend to avoid Docker.

**Comment by sademakn** [comment]
> Note that you in addition to checking how much memory your (bare metal) machine has, you should also check how much of that memory is accessible by the Docker container for use as shared memory, which is typically a much smaller amount. For reasons such as these, we recommend to avoid Docker.

I fixed my problem with increasing shm-size in docker-compose like this :
```
version: "3.7"

services:
  pgb_app:
    container_name: pgb_app
    build: .
    shm_size: '20gb'

```
